### PR TITLE
dstore: fix errors in offset calculation

### PR DESCRIPTION
**Pull request by hjelmn**
This commit fixes two errors in the bounds checking in
put_data_to_the_end.

Signed-off-by: Nathan Hjelm <hjelmn@lanl.gov>

### CODE REVIEW

**Comment by hjelmn** [comment]
@artpol84 Hit the problem again @ 2k nodes. The error is pretty self-evident. Even with this commit there is still a segment overflow and I don't have time to dig into it now. Adding a small constant to the left hand side of the bounds check made it work. I am running with:

```
if ((offset + 8 + ESH_KEY_SIZE(key, size) + EXT_SLOT_SIZE()) > _data_segment_size)  {
```

So it looks like something in ESH_KEY_SIZE is wrong? I can look further tomorrow if the error hasn't been found.

File: src/dstore/pmix_esh.c [hunk 1]
```diff
@@ -2126,15 +2128,15 @@ static size_t put_data_to_the_end(ns_track_elem_t *ns_info, seg_desc_t *dataseg,
...
+    if ((sizeof(size_t) + ESH_KEY_SIZE(key, size) + EXT_SLOT_SIZE(key)) > _data_segment_size) {
         /* this is an error case: segment is so small that cannot place evem a single key-value pair.
          * warn a user about it and fail. */
         offset = 0; /* offset cannot be 0 in normal case, so we use this value to indicate a problem. */
         pmix_output(0, "PLEASE set NS_DATA_SEG_SIZE to value which is larger when %lu.",
                 sizeof(size_t) + strlen(key) + 1 + sizeof(size_t) + size + EXT_SLOT_SIZE(key));
         return offset;
     }
-    if (offset + ESH_KEY_SIZE(key, size) + EXT_SLOT_SIZE(key) > _data_segment_size)  {
+    if ((offset + ESH_KEY_SIZE(key, size) + EXT_SLOT_SIZE(key)) > _data_segment_size)  {
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: * ` EXT_SLOT_SIZE(key)` - to make sure: is this fixed in your version? should be like this:
https://github.com/pmix/master/pull/282/files#diff-ef98725cc50c8a568d8e5622d505e776R2137

File: src/dstore/pmix_esh.c [hunk 2]
```diff
@@ -2107,9 +2110,8 @@ static int put_empty_ext_slot(seg_desc_t *dataseg)
 
 static size_t put_data_to_the_end(ns_track_elem_t *ns_info, seg_desc_t *dataseg, char *key, void *buffer, size_t size)
 {
-    size_t offset;
+    size_t offset, id = 0;
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: so you say that integer was overflowed?
segment size if 4K by default. To overflow it you need to store 2G*4K which is unlikely I hope.

File: src/dstore/pmix_esh.c [hunk 3]
```diff
@@ -2126,15 +2128,15 @@ static size_t put_data_to_the_end(ns_track_elem_t *ns_info, seg_desc_t *dataseg,
     offset = global_offset % _data_segment_size;
 
     /* We should provide additional space at the end of segment to place EXTENSION_SLOT to have an ability to enlarge data for this rank.*/
-    if (sizeof(size_t) + ESH_KEY_SIZE(key, size) + EXT_SLOT_SIZE(key) > _data_segment_size) {
+    if ((sizeof(size_t) + ESH_KEY_SIZE(key, size) + EXT_SLOT_SIZE(key)) > _data_segment_size) {
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: According to this:
http://en.cppreference.com/w/c/language/operator_precedence
this line was fine.
can you explain what two errors you identified?

File: src/dstore/pmix_esh.c [hunk 4]
```diff
@@ -2126,15 +2128,15 @@ static size_t put_data_to_the_end(ns_track_elem_t *ns_info, seg_desc_t *dataseg,
...
         /* this is an error case: segment is so small that cannot place evem a single key-value pair.
          * warn a user about it and fail. */
         offset = 0; /* offset cannot be 0 in normal case, so we use this value to indicate a problem. */
         pmix_output(0, "PLEASE set NS_DATA_SEG_SIZE to value which is larger when %lu.",
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: according to the error you previously reported you should either see this message

**Comment by artpol84** [comment]
@hjelmn can you also remind me what version of PMIx you are using? We need to know this because ESH code diverged in master so it's important to know.

File: src/dstore/pmix_esh.c (hunk 1 above)
Comment (artpol84) [inline src/dstore/pmix_esh.c]: I see that you are using updated version of `EXT_SLOT_SIZE`

File: src/dstore/pmix_esh.c (hunk 2 above)
Comment (artpol84) [inline src/dstore/pmix_esh.c]: What I'm trying to say is that I don't mind changing the type of  `id` but I don't believe it was the reason of the error.

**Comment by artpol84** [comment]
@hjelmn And I'll duplicate my question here: what error you hit this time:
* This https://github.com/pmix/master/pull/282#issuecomment-277422536
* or this https://github.com/pmix/master/pull/282#issuecomment-277454198

**Comment by jjhursey** [comment]
@hjelmn After we merge in PR #284 you will need to rebase this branch so we can pick up the `int` to `size_t` change.

**Comment by jjhursey** [comment]
I just noticed this was a PR against `master` (updated the milestone). PR #282 will merge in, then you'll need to rebase. If you want this change in `v1.2` then you can cherry-pick it over in a new PR. Thanks!

**Pull request closed without merging**